pkg/collection/ducks.go

```diff
@@ -0,0 +1,200 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collection
+
+import (
+	"sort"
+	"strings"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"knative.dev/discovery/pkg/apis/discovery/v1alpha1"
+)
+
+type DuckHunter interface {
+	AddCRDs(crds []*apiextensionsv1.CustomResourceDefinition)
+	AddCRD(crd *apiextensionsv1.CustomResourceDefinition)
+	// AddRef is used to insert built-in resource types.
+	// TODO: if the ref is a CRD, load the CRD and pass that CRD to AddCRD.
+	AddRef(duckVersion string, ref v1alpha1.ResourceRef)
+	Ducks() map[string][]v1alpha1.ResourceMeta
+}
+
+type DuckFilters struct {
+	// DuckLabel is expected to be in the form:
+	//	 `<group>/<names.singular>`
+	DuckLabel string
+	// DuckVersionFormat is expected to be in the form:
+	//   `<names.plural>.<group>`
+	// and will be used to assemble `<names.plural>.<group>/<duckVersion>`.
+	// This will group duck type versions based on the given annotation base.
+	DuckVersionPrefix string
+}
+
+// NewDuckHunter
+// versions are used to default all the DuckType versions that apply to unfiltered CRDs.
+func NewDuckHunter(versions []v1alpha1.DuckVersion, filters *DuckFilters) DuckHunter {
+	dh := &duckHunter{
+		filters:  filters,
+		versions: make([]string, 0),
+		ducks:    make(map[string][]v1alpha1.ResourceMeta, len(versions)),
+	}
+
+	for _, v := range versions {
+		if _, found := dh.ducks[v.Name]; !found {
+			dh.versions = append(dh.versions, v.Name)
+			dh.ducks[v.Name] = make([]v1alpha1.ResourceMeta, 0)
+		}
+	}
+
+	return dh
+}
+
+type duckHunter struct {
+	filters *DuckFilters
+	// versions are the versions to use for unfiltered CRDs being added to the ducks map.
+	versions []string
+	ducks    map[string][]v1alpha1.ResourceMeta
+}
+
+func (dh *duckHunter) AddCRDs(crds []*apiextensionsv1.CustomResourceDefinition) {
+	for _, crd := range crds {
+		dh.AddCRD(crd)
+	}
+}
+
+func (dh *duckHunter) AddCRD(crd *apiextensionsv1.CustomResourceDefinition) {
+	if crd == nil {
+		return
+	}
+	if metas := crdToResourceMeta(crd); len(metas) > 0 {
+		dh.collectVersionsByFilter(crd)
+		for _, meta := range metas {
+			if !dh.handledWithFilters(crd, meta) {
+				// TODO: here is where the label filters would be tested for duck version annotation matching.
+				for _, v := range dh.versions {
+					dh.ducks[v] = append(dh.ducks[v], meta)
+				}
+			}
+		}
+	}
+}
+
+func (dh *duckHunter) handledWithFilters(crd *apiextensionsv1.CustomResourceDefinition, meta v1alpha1.ResourceMeta) bool {
```

Comment:
maybe `*meta v1alpha1.ResourceMeta` ?

Reply:
I did not just because it is a small resource and I don't mind it being passed immutably is the thinking here